cloud/pkg/cloudhub/dispatcher/message_dispatcher.go

```diff
@@ -0,0 +1,438 @@
+	msg.SetResourceOperation(fmt.Sprintf("node/%s/%s", info.NodeID, msg.GetResource()), msg.GetOperation())
+	if model.IsFromEdge(msg) {
+		return md.Publish(msg)
+	}
+	return nil
+}
+
+func (md *messageDispatcher) enqueueNoAckMessage(nodeID string, msg *beehivemodel.Message) {
+	nodeMessagePool := md.GetNodeMessagePool(nodeID)
+
+	messageKey, _ := cloudhubutil.NoAckMessageKeyFunc(msg)
+	if err := nodeMessagePool.NoAckMessageStore.Add(msg); err != nil {
+		klog.Errorf("failed to add msg: %v", err)
+		return
+	}
+	nodeMessagePool.NoAckMessageQueue.Add(messageKey)
+}
+
+func (md *messageDispatcher) enqueueAckMessage(nodeID string, msg *beehivemodel.Message) {
+	// Message that require ack MUST have resource version.
+	if msg.GetResourceVersion() == "" && !isDeleteMessage(msg) {
+		return
+	}
+
+	nodeMessagePool := md.GetNodeMessagePool(nodeID)
+	nodeQueue := nodeMessagePool.AckMessageQueue
+	nodeStore := nodeMessagePool.AckMessageStore
+
+	messageKey, err := cloudhubutil.AckMessageKeyFunc(msg)
+	if err != nil {
+		klog.Errorf("fail to get key for message: %s", msg.String())
+		return
+	}
+
+	shouldEnqueue := false
+	defer func() {
+		if shouldEnqueue {
+			if err := nodeStore.Add(msg); err != nil {
+				klog.Errorf("fail to add message %v nodeStore, err: %v", msg, err)
+				return
+			}
+			nodeQueue.Add(messageKey)
+		}
+	}()
+
+	// If the message operation is delete, force to sync the resource message
+	// If the message operation is response, force to sync the resource message,
+	// since the edgeCore requests it.
+	if isDeleteMessage(msg) || msg.GetOperation() == beehivemodel.ResponseOperation {
+		shouldEnqueue = true
+		return
+	}
+
+	item, exist, _ := nodeStore.GetByKey(messageKey)
+	if exist {
+		msgInStore := item.(*beehivemodel.Message)
+
+		if isDeleteMessage(msgInStore) ||
+			synccontroller.CompareResourceVersion(msg.GetResourceVersion(), msgInStore.GetResourceVersion()) <= 0 {
+			// If the message resource version is older than the message in store or the operation
+			// for the message in store is delete, The message will be discarded directly.
+			return
+		}
+		shouldEnqueue = true
+		return
+	}
+
+	// If the message doesn't exist in the store, then compare it with the version stored in the objectSync.
+	resourceNamespace, _ := messagelayer.GetNamespace(*msg)
+	resourceName, _ := messagelayer.GetResourceName(*msg)
+	resourceUID, err := cloudhubutil.GetMessageUID(*msg)
+	if err != nil {
+		klog.Errorf("fail to get message UID for message: %s", msg.Header.ID)
+		return
+	}
+
+	objectSyncName := synccontroller.BuildObjectSyncName(nodeID, resourceUID)
+	objectSync, err := md.objectSyncLister.ObjectSyncs(resourceNamespace).Get(objectSyncName)
+
+	switch {
+	case err == nil && objectSync.Status.ObjectResourceVersion != "":
+		if synccontroller.CompareResourceVersion(msg.GetResourceVersion(), objectSync.Status.ObjectResourceVersion) > 0 {
+			shouldEnqueue = true
+			return
+		}
+
+	case err != nil && apierrors.IsNotFound(err):
```

Comment:
`ObjectSyncs` represent current resource real status on the Edge node, I think we should only create `ObjectSyncs` after we receive the ack message.

Reply:
according to the issue
https://github.com/kubeedge/kubeedge/issues/3243
https://github.com/kubeedge/kubeedge/pull/3244

we need this for the message to avoid that the message is lost when the edge node reconnect or cloudcore reboot